crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -190,16 +190,13 @@ impl Format<PyFormatContext<'_>> for MaybeParenthesizeExpression<'_> {
             return expression.format().with_options(Parentheses::Always).fmt(f);
         }
 
-        let needs_parentheses = expression.needs_parentheses(*parent, f.context());
-        let needs_parentheses = match parenthesize {
-            Parenthesize::IfRequired => match needs_parentheses {
-                OptionalParentheses::Always => OptionalParentheses::Always,
-                _ if f.context().node_level().is_parenthesized() => OptionalParentheses::Never,
-                needs_parentheses => needs_parentheses,
-            },
-            Parenthesize::Optional
-            | Parenthesize::IfBreaks
-            | Parenthesize::IfBreaksOrIfRequired => needs_parentheses,
+        let needs_parentheses = match expression.needs_parentheses(*parent, f.context()) {
```

Comment:
Applying this logic to all parenthesized cases reduces the overall number of branches.